sophia-oracle-introduction.md

```diff
@@ -26,128 +28,157 @@ contract Greeter =
+First thing to notice is that the function has an `entrypoint` keyword. This means that it can be called outside the contract. The function is also `stateful`. Any function that modifies the contract's state **must** be declared as `stateful`. And lastly, the function is payable. A payable function will allow tokens to be sent to the contract when it is called. The caller of the function is expected to send 10 ae tokens when calling the function.
+
+We're going to be passing our message as an argument of type `string`, as expected by the oracle. Now, there are a few important things to look at in the following lines. Firstly, we're going to be making a query to the oracle that we registered previously and store the reference to that query in a variable. The data of that query will of course be the message that we pass as an argument. We're also going to respect the fee required by our oracle. The two TTLs that are being passed respectively are for the time (in key blocks) during which our query object will be available for an answer and the time during which the response will be retrievable from the blockchain's state tree.
 
 After that, we are calling the builtin function `put`, which modifies the state of the contract. Since our state is a `record`, we are going to use the update syntax for records and maps to update it. `state{greets @ g = (caller_address, query) :: g}` translates to: the field `greets` of the record `state` will now have a value of `(caller_address, query) :: g`. The odd thing here is the `@` syntax. This is a way of binding the current value of the field before the `@` to the name provided after it. It could also be written as `state{greets = (caller_address, query) :: state.greets}`. As for `(caller_address, query) :: g`, this is the syntax for prepending a new element to a list. If we look at the initial definition of our state, we said that it will keep track of queries and addresses that make them in a list. So, we take the `query`, pair it with the string `caller address` and add it to the current list that is in the state.
 
 ## Responding to the greetings
+
 Now that we are able to query our oracle, we should also add logic to respond to those queries. We will have to work with a recursive function that iterates through all of the queries and responds to them. To make it a bit more complicated, we will make it so that the oracle responds only to certain greets, let's say "Hello", "Hi" and "Greetings".
 
 Let's begin with the public function that will start the recursive calls:
-```
-public function respond_to_greets() =
-    respond_to_greets'(state.greets)
+
+```sophia
+stateful entrypoint respond_to_greets() =
+  respond_to_greets'(state.greets)
```

Comment:
```suggestion
  [ respond_to_greet(caller_addr, query) | (caller_addr, query) <- state.greets) ]
```

Reply:
Will simplify the code and you can skip the explanation about not being able to have to functions with the same name below.

Reply:
Thanks for the feedback @hanssv

` [ respond_to_greet(caller_addr, query) | (caller_addr, query) <- state.greets) ]`

a similar thing [respond_to_greet(c,q) for c,q in state[greets]] is called list cmprehension in python. What is this called in sophia?

Reply:
The concept in general is called list comprehension. https://github.com/aeternity/protocol/blob/master/contracts/sophia.md#lists